This notebook will use the tools we have created to see what the optimal counter-army is to the Terran Bio build, which revolves around biological units supported by Medivacs. The biological Terran units include the Marine, Marauder, Reaper, and Ghost. Reapers and Ghosts are mainly used for harassment and not large scale battles. A typical Terran Bio build has Marines and Marauders with a ratio around 3:1, and Medivacs with a ratio around 1 Medivac for every 10 or so infantry units. Two common ratios of Marine:Marauder:Medivac are 6:2:1 and 15:5:2. For our purposes, we will be testing the 15:5:2 ratio and optimizing against 45 Marines, 15 Marauders, and 6 Medivacs. This army has a total supply of 87, simulating an early- to mid-game composition. By default, our random army generator creates armies with total supply ranging anywhere between 0 and 200. To reduce the amount of iterations needed to hone in on the optimal solution, we will be limiting our random armies to a supply cap of 120, since we know for certain that pretty much any army with a supply greater than that will be suboptimal (such armies will likely be overkill, or contain too many support units that do not add enough value to the overall army composition).

The Terran Bio build is standard because it revolves around Marines which are cheap, one of the first units Terran players can produce, and are extremely versatile as they are fast, ranged, and can attack both ground and air units. Marauders act as meat-shields in addition to extra damage against armored units, and Medivacs provide constant healing to both Marines and Marauders. While the functionality is not included in our combat simulator, Medivacs can also carry a small group of Marines and quickly fly them around the map, making this build useful for harassment in addition to all-out battle.

In [57]:
from sim_units import get_Units, get_Terran, get_Protoss, get_Zerg
from generate_comps import get_army_supply, init_army_comps
from linear_program import find_optimal_army, get_army_cost
from viability import viability
import json
from joblib import Parallel, delayed
import time

In [52]:
def find_counters_parallel(army_comp, file_name, race, supply_cap=200, num_tests=10, test_comps=None):
    """
    army_comp is the enemy army composition we are optimizing against
    file_name is a string, name of txt file we will save our results to
    race:{'Terran', 'Protoss', 'Zerg'} the race we are optimizing for
    supply_cap is an integer of the largest army we wish to test
    num_tests is an integer, the number of times we wish to run our test
    test_comps is a list of army compositions. Optional parameter in case we
    want to use that specific list of armies instead of random ones
    Creates a file with the army counters
    """
    army_list = []
    for i in range(num_tests):
        army_list.append(army_comp)
    start_time = time.time()
    terran_counters = {}
    results = Parallel(n_jobs=-1)(delayed(find_optimal_army)(enemy, race, supply_cap=supply_cap,
                                                             test_comps=test_comps)
                             for enemy in army_list)
    run_time = time.time() - start_time
    print(run_time)
    with open(file_name, 'a') as file:
        file.write("Enemy army:")
        file.write('\n')
        file.write(str(army_comp))
        file.write('\n')
        file.write('\n')
        file.write(str(run_time))
        file.write('\n')
        file.write('\n')
        file.write("Counters:")
        for line in results:
            file.write(line)
            file.write('\n')
        file.write('\n')
        file.write('\n')
        file.write('\n')
    return results

In [3]:
terran_bio = {}
terran_bio['Marine'] = 45
terran_bio['Marauder'] = 15
terran_bio['Medivac'] = 6

In [ ]:
find_counters_parallel(terran_bio, 'terran_bio.txt', 'Terran', supply_cap=125, min_supply=0)

In [ ]:
for i in range(10):
    find_counters_parallel(terran_bio, 'terran_bio.txt', 'Terran', supply_cap=125, min_supply=0)

The following is the output taken from terran_bio.txt. We can see that

In [28]:
bio_counters = []

In [29]:
c1 = {}
c1['Marine'] = 122

c2 = {}
c2['Hellbat'] = 20
c2['Viking'] = 27

c3 = {}
c3['SiegeTank'] = 30

c4 = {}
c4['SiegeTank'] = 26

c5 = {}
c5['Cyclone'] = 31

c6 = {}
c6['SiegeTank'] = 26

c7 = {}
c7['Thor'] = 17
c7['Viking'] = 1

c8 = {}
c8['Hellbat'] = 53

c9 = {}
c9['Marine'] = 46
c9['Hellion'] = 14
c9['Liberator'] = 2

c10 = {}
c10['Hellion'] = 48

In [30]:
bio_counters.append(c1)
bio_counters.append(c2)
bio_counters.append(c3)
bio_counters.append(c4)
bio_counters.append(c5)
bio_counters.append(c6)
bio_counters.append(c7)
bio_counters.append(c8)
bio_counters.append(c9)
bio_counters.append(c10)

In [31]:
c11 = {}
c11['Marauder'] = 22

c12 = {}
c12['Thor'] = 2

c13 = {}
c13['SiegeTank'] = 25

c14 = {}
c14['Hellbat'] = 30

c15 = {}
c15['Hellion'] = 21
c15['Marauder'] = 1
c15['Reaper'] = 58

c16 = {}
c16['Ghost'] = 40

c17 = {}
c17['Reaper'] = 71
c17['Cyclone'] = 10

c18 = {}
c18['Medivac'] = 5
c18['Banshee'] = 24

c19 = {}
c19['Reaper'] = 104

c20 = {}
c20['Reaper'] = 115

In [32]:
bio_counters.append(c11)
bio_counters.append(c12)
bio_counters.append(c13)
bio_counters.append(c14)
bio_counters.append(c15)
bio_counters.append(c16)
bio_counters.append(c17)
bio_counters.append(c18)
bio_counters.append(c19)
bio_counters.append(c20)

In [33]:
c21 = {}
c21['Hellion'] = 7
c21['Marauder'] = 33
c21['Battlecruiser'] = 2

c22 = {}
c22['Hellion'] = 27
c22['Thor'] = 8

c23 = {}
c23['Liberator'] = 28

c24 = {}
c24['Marauder'] = 28
c24['Banshee'] = 7

c25 = {}
c25['Marauder'] = 52

c26 = {}
c26['Banshee'] = 35

c27 = {}
c27['Medivac'] = 20
c27['Banshee'] = 23

c28 = {}
c28['SiegeTank'] = 25

c29 = {}
c29['Hellion'] = 30
c29['Ghost'] = 16

c30 = {}
c30['Ghost'] = 21
c30['Hellbat'] = 20

In [34]:
bio_counters.append(c21)
bio_counters.append(c22)
bio_counters.append(c23)
bio_counters.append(c24)
bio_counters.append(c25)
bio_counters.append(c26)
bio_counters.append(c27)
bio_counters.append(c28)
bio_counters.append(c29)
bio_counters.append(c30)

In [35]:
c31 = {}
c31['Marauder'] = 48
c31['Banshee'] = 1

c32 = {}
c32['Hellbat'] = 46

c33 = {}
c33['Liberator'] = 9
c33['SiegeTank'] = 20

c34 = {}
c34['Ghost'] = 30
c34['Banshee'] = 14

c35 = {}
c35['Liberator'] = 37

c36 = {}
c36['Liberator'] = 24

c37 = {}
c37['SiegeTank'] = 20
c37['Thor'] = 3

c38 = {}
c38['Reaper'] = 42
c38['Liberator'] = 18

c39 = {}
c39['Marauder'] = 54

c40 = {}
c40['Medivac'] = 19
c40['Reaper'] = 73

In [36]:
bio_counters.append(c31)
bio_counters.append(c32)
bio_counters.append(c33)
bio_counters.append(c34)
bio_counters.append(c35)
bio_counters.append(c36)
bio_counters.append(c37)
bio_counters.append(c38)
bio_counters.append(c39)
bio_counters.append(c40)

In [37]:
c41 = {}
c41['Cyclone'] = 38

c42 = {}
c42['Marine'] = 53
c42['Hellion'] = 9

c43 = {}
c43['Reaper'] = 57
c43['SiegeTank'] = 5

c44 = {}
c44['Hellbat'] = 42

c45 = {}
c45['Banshee'] = 40

c46 = {}
c46['Hellion'] = 19
c46['Battlecruiser'] = 3

c47 = {}
c47['Medivac'] = 10
c47['Hellbat'] = 35

c48 = {}
c48['Marine'] = 26
c48['Hellbat'] = 37

c49 = {}
c49['Hellion'] = 43

c50 = {}
c50['SiegeTank'] = 26
c50['Battlecruiser'] = 7

In [38]:
bio_counters.append(c41)
bio_counters.append(c42)
bio_counters.append(c43)
bio_counters.append(c44)
bio_counters.append(c45)
bio_counters.append(c46)
bio_counters.append(c47)
bio_counters.append(c48)
bio_counters.append(c49)
bio_counters.append(c50)

In [39]:
c51 = {}
c51['SiegeTank'] = 23
c51['Viking'] = 27

c52 = {}
c52['Ghost'] = 25
c52['Hellbat'] = 18

c53 = {}
c53['Hellbat'] = 41

c54 = {}
c54['Marine'] = 78

c55 = {}
c55['Reaper'] = 111

c56 = {}
c56['Hellion'] = 25
c56['Cyclone'] = 6
c56['Banshee'] = 6

c57 = {}
c57['Ghost'] = 62

c58 = {}
c58['Hellbat'] = 5
c58['Battlecruiser'] = 9

c59 = {}
c59['Hellion'] = 20
c59['Reaper'] = 30
c59['Hellbat'] = 5

c60 = {}
c60['Medivac'] = 4
c60['SiegeTank'] = 2
c60['Battlecruiser'] = 6

In [40]:
bio_counters.append(c51)
bio_counters.append(c52)
bio_counters.append(c53)
bio_counters.append(c54)
bio_counters.append(c55)
bio_counters.append(c56)
bio_counters.append(c57)
bio_counters.append(c58)
bio_counters.append(c59)
bio_counters.append(c60)

In [41]:
c61 = {}
c61['Hellbat'] = 29
c61['Liberator'] = 10

c62 = {}
c62['Reaper'] = 33
c62['Banshee'] = 19

c63 = {}
c63['Cyclone'] = 39

c64 = {}
c64['Hellbat'] = 2
c64['Thor'] = 20

c65 = {}
c65['Hellbat'] = 18
c65['Battlecruiser'] = 11

c66 = {}
c66['Banshee'] = 34

c67 = {}
c67['Liberator'] = 21

c68 = {}
c68['SiegeTank'] = 35

c69 = {}
c69['Reaper'] = 96

c70 = {}
c70['Hellbat'] = 60

In [42]:
bio_counters.append(c61)
bio_counters.append(c62)
bio_counters.append(c63)
bio_counters.append(c64)
bio_counters.append(c65)
bio_counters.append(c66)
bio_counters.append(c67)
bio_counters.append(c68)
bio_counters.append(c69)
bio_counters.append(c70)

In [43]:
c71 = {}
c71['Marauder'] = 61

c72 = {}
c72['SiegeTank'] = 34

c73 = {}
c73['Banshee'] = 32

c74 = {}
c74['Liberator'] = 30

c75 = {}
c75['Cyclone'] = 32
c75['Banshee'] = 6

c76 = {}
c76['Reaper'] = 69
c76['Hellbat'] = 11

c77 = {}
c77['Ghost'] = 31
c77['Hellbat'] = 14

c78 = {}
c78['Marauder'] = 24
c78['Liberator'] = 11

c79 = {}
c79['Hellbat'] = 36

c80 = {}
c80['Reaper'] = 51
c80['Hellbat'] = 31

In [44]:
bio_counters.append(c71)
bio_counters.append(c72)
bio_counters.append(c73)
bio_counters.append(c74)
bio_counters.append(c75)
bio_counters.append(c76)
bio_counters.append(c77)
bio_counters.append(c78)
bio_counters.append(c79)
bio_counters.append(c80)

In [45]:
c81 = {}
c81['Hellion'] = 25
c81['Ghost'] = 26

c82 = {}
c82['Thor'] = 12

c83 = {}
c83['Thor'] = 2
c83['Viking'] = 5
c83['Banshee'] = 24

c84 = {}
c84['Hellion'] = 49

c85 = {}
c85['Marine'] = 34
c85['Hellbat'] = 44

c86 = {}
c86['Hellbat'] = 47

c87 = {}
c87['Marauder'] = 62

c88 = {}
c88['Medivac'] = 6
c88['Viking'] = 5
c88['Liberator'] = 16

c89 = {}
c89['SiegeTank'] = 4
c89['Cyclone'] = 30

c90 = {}
c90['Liberator'] = 19

In [46]:
bio_counters.append(c81)
bio_counters.append(c82)
bio_counters.append(c83)
bio_counters.append(c84)
bio_counters.append(c85)
bio_counters.append(c86)
bio_counters.append(c87)
bio_counters.append(c88)
bio_counters.append(c89)
bio_counters.append(c90)

In [47]:
c91 = {}
c91['Hellbat'] = 31
c91['Thor'] = 7

c92 = {}
c92['Hellion'] = 42

c93 = {}
c93['Marine'] = 96
c93['Viking'] = 9

c94 = {}
c94['Ghost'] = 27
c94['Cyclone'] = 17

c95 = {}
c95['Battlecruiser'] = 7
c95['Cyclone'] = 16

c96 = {}
c96['Viking'] = 17
c96['Battlecruiser'] = 8

c97 = {}
c97['Hellion'] = 50
c97['Liberator'] = 1

c98 = {}
c98['Cyclone'] = 36
c98['Viking'] = 4

c99 = {}
c99['Hellion'] = 9
c99['Marauder'] = 27

c100 = {}
c100['Cyclone'] = 17
c100['Liberator'] = 19

In [48]:
bio_counters.append(c91)
bio_counters.append(c92)
bio_counters.append(c93)
bio_counters.append(c94)
bio_counters.append(c95)
bio_counters.append(c96)
bio_counters.append(c97)
bio_counters.append(c98)
bio_counters.append(c99)
bio_counters.append(c100)

In [49]:
c101 = {}
c101['Ghost'] = 14
c101['Hellbat'] = 21

c102 = {}
c102['Hellion'] = 4
c102['Liberator'] = 17

c103 = {}
c103['Marauder'] = 52

c104 = {}
c104['Reaper'] = 38
c104['Ghost'] = 22

c105 = {}
c105['Marine'] = 100
c105['Banshee'] = 3

c106 = {}
c106['Hellbat'] = 49

c107 = {}
c107['Hellion'] = 40

c108 = {}
c108['Viking'] = 8
c108['Liberator'] = 16

c109 = {}
c109['Banshee'] = 24
c109['Battlecruiser'] = 5

c110 = {}
c110['Hellion'] = 43

In [50]:
bio_counters.append(c101)
bio_counters.append(c102)
bio_counters.append(c103)
bio_counters.append(c104)
bio_counters.append(c105)
bio_counters.append(c106)
bio_counters.append(c107)
bio_counters.append(c108)
bio_counters.append(c109)
bio_counters.append(c110)

Having run our program against random enemy armies 110 times

In [53]:
find_counters_parallel(terran_bio, 'terran_bio_final.txt', 'Terran', test_comps=bio_counters)

2353.6373023986816


["Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units",
 "Army Comp{'Cyclone': 32, 'Banshee': 6}, Terran= 0.0, viability:units"]

In [54]:
get_army_supply(c75)

114

In [55]:
get_army_cost(c75)

9500

In [56]:
Units = get_Units()
minerals = Units['Cyclone']['mineral'] * 32 + Units['Banshee']['mineral'] * 6
gas = minerals = Units['Cyclone']['gas'] * 32 + Units['Banshee']['gas'] * 6
print("minerals: " + str(minerals))
print("gas: " + str(gas))

minerals: 3800
gas: 3800


From here we can see that the most optimal composition our program came up with was the 75th output of 32 Cyclones and 6 Banshees, for a total supply of 114 and a cost of 3800 minerals and 3800 gas. However, this is still not the most optimal composition. We can quickly run a test with slightly lower unit numbers and see that this cheaper army composition is still viable. 

In [58]:
cheaper_comp1 = {}
cheaper_comp1['Cyclone'] = 30
cheaper_comp1['Banshee'] = 5

In [59]:
viability(terran_bio, cheaper_comp1)

1

In [60]:
cheaper_comp2 = {}
cheaper_comp2['Cyclone'] = 30
cheaper_comp2['Banshee'] = 0

In [61]:
viability(terran_bio, cheaper_comp2)

1